### PR TITLE
Fix array handling, fix tests, other smaller fixes

### PR DESCRIPTION
**Pull request by GUI**
I have a number of commits here to fix a few issues I encountered and also resurrect the test suite for this project. Sorry to pile all this into a single pull request, but hopefully each commit is pretty explanatory in what it fixes. I could try to pull them apart into separate pull requests, but they all sort of become part of a larger effort to get the entire test suite passing again (I had started with just wanting to fix issues with arrays, but I wanted to get some sort of verification I wasn't breaking other things, so that led me down this testing rabbit-hole).

In any case, here's a summary of what's changed and fixed:
- Fixes wonkiness with array storage and fetching, related to this issue: https://github.com/bigplum/lua-resty-mongol/issues/29 (I'd consider the previous behavior broken, but see my comment for more detail the behavior change: https://github.com/GUI/resty-mongol/commit/9c6d2c5b8d099670c0f5e0cd87d77d0305e05e43)
- Fixes PID handling in OS X: https://github.com/bigplum/lua-resty-mongol/pull/25 Also improves efficiency and allows the module to be used inside `init_worker_by_lua` by using `ngx.worker.pid()`.
- Fixes issues with `cursor:limit()` that the existing tests uncovered (in some cases, the limit wasn't being applied and the cursor would iterate over all the results, despite having a limit).
- Makes some minor fixes in edge-cases around gridfs and authentication that the existing tests uncovered.
- I synced up the test suite with the more recent additions to [bigplum/lua-resty-mongol](https://github.com/bigplum/lua-resty-mongol/commits/master).
- I added a `make test` command that should do everything necessary to run the tests.
- I added configuration for CircleCI, so you can have the tests automatically run there, where all tests are now passing: https://circleci.com/gh/GUI/resty-mongol/9
- Updated the text in the README to take into account this module's different name (`resty-mongol` instead of `resty.mongol`) and it's different style of constructor for the connection.

Let me know if you have any questions/comments/concerns about any of this. Thanks!

### CODE REVIEW

**Comment by DorianGray** [comment]
This looks super solid. Thanks so much! I'll review a little closer, make a few additions, and merge shortly.

**Comment by DorianGray** [comment]
Funny, because this was literally my next goal.

**Comment by DorianGray** [comment]
@GUI I've tested out these changes. They look good, with the exception of them breaking skip. I've marked what I believe is the offending line, would you mind looking at a different way to solve that particular problem? Skip was never implemented properly in bigplum's version, I implemented it here but it did have some issues(that were able to be worked around).

File: src/cursor.lua [hunk 1]
```diff
@@ -59,7 +59,7 @@ function cursor_methods:count()
 end
 
 function cursor_methods:next()
-  if self.limit_n > 0 and self.i - self.skip_n >= self.limit_n then return nil end
+  if self.limit_n > 0 and self.i >= self.limit_n then return nil end
```
Comment (DorianGray) [inline src/cursor.lua]: I believe this breaks skip.

Comment (DorianGray) [inline src/cursor.lua]: To be explicit, we don't want to keep -all- records in memory, we only want to keep this set of records in memory. We return a generator, not a list.

**Comment by GUI** [comment]
I believe `skip` is now fixed in a way that also fixes the original `limit` batching issue with the changes in: https://github.com/GUI/resty-mongol/commit/f1575798dc99b2dc4d59cb90804dfc7db5cd3121 I also added tests around the skip function. Did you want to take another look to see if this makes sense and appears correct to you? To fix this, I added an `overall_index` variable to keep track of where the cursor is in the complete result set (regardless of batches), whereas the `i` variable now just keeps track of where the cursor is inside the single batch of results that's currently in memory.

I also uncovered an issue with the `t/objid.t` test and fixed that in https://github.com/GUI/resty-mongol/commit/553cde4e3ceb9390fcb1ba72d4bae933802d5008.

**Comment by johnae** [comment]
@DorianGray @GUI would be great if this was released. That mongo:new() vs mongo() really had me confused for a long time - thanks for documenting this somewhat different usage.
